File: src/solutions/2025/01/python/main.py
import sys
import os
import typing

sys.path.append(os.getcwd())
from src.langs.python.run import run


def main() -> None:
   options = {'has_alternate': False, 'has_io': False}
   run(sys.argv, part_one, part_two, options)


def solve(input: str, p2: bool) -> str:
   dial = 50
   zero = 0
   for n in input.split('\n'):
      newDial = dial + int(n[1:]) * (1 if n[0] == 'R' else -1)

      if p2:
         zero += abs(newDial) // 100 + (dial and newDial <= 0)
         dial = newDial % 100
      else:
         zero += not (dial := newDial % 100)

   return str(zero)


def part_one(input: str, is_test: bool) -> str:
   return solve(input, False)


def part_two(input: str, is_test: bool) -> str:
   return solve(input, True)


if __name__ == '__main__':
   main()

File: src/solutions/2025/01/c/main.c
#include "main.h"
#include "run.h"
#include "utils_math.h"
#include "utils_num.h"
#include "utils_str.h"
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static char *solve(const char *restrict input, const int32_t isTest,
                   const int32_t p2) {
   size_t size;
   char **lines = str_splitc(input, '\n', &size);

   int32_t dial = 50;
   int32_t zero = 0;
   for (size_t i = 0; i < size; i++) {
      char *line = lines[i];
      char moveStr[3] = {0};
      str_slice(line, moveStr, 1, strlen(line));
      int32_t newDial = dial + atoi(moveStr) * (line[0] == 'R' ? 1 : -1);

      if (p2) {
         zero += abs(newDial) / 100 + (dial != 0 && newDial <= 0 ? 1 : 0);
         dial = math_i32mod(newDial, 100);
      } else {
         dial = newDial;
         if (dial % 100 == 0) {
            zero++;
         }
      }

      free(line);
   }

   free(lines);
   return num_tostr(zero);
}

static char *part1(const char *restrict input, const int32_t isTest) {
   return solve(input, isTest, 0);
}

static char *part2(const char *restrict input, const int32_t isTest) {
   return solve(input, isTest, 1);
}

int main(int argc, char *argv[]) {
   solution_options_t options = getoptions();
   return run(argc, argv, part1, part2, options);
}
